File: 0059-spiral-matrix-ii/0059-spiral-matrix-ii.php
class Solution {

    /**
     * @param Integer $n
     * @return Integer[][]
     */
    function generateMatrix($n) {
        $matrix=array();
        for ($i=0; $i <$n ; $i++) { 
            for ($j=0; $j < $n; $j++) { 
                $matrix[$i][$j]=0;
            }
        }
        $i=1;
        $top=0;
        $bottom=$n-1;
        $left=0;
        $ritht=$n-1;
        $direction=0;
        while ($i<=$n*$n) 
        {
            if ($direction==0) 
            {
                for ($j=$left; $j <=$ritht ; $j++) 
                { 
                    $matrix[$top][$j]=$i++;
                    
                }
                $top++;
            }
            elseif ($direction==1) 
            {
                for ($j=$top; $j <=$bottom ; $j++) 
                { 
                    $matrix[$j][$ritht]=$i++;
                }
                $ritht--;
            }
            elseif ($direction==2) {
            for ($j=$ritht; $j >=$left ; $j--) { 
                $matrix[$bottom][$j]=$i++;
            }
            $bottom--;
            }
            else {
            for ($j=$bottom; $j >=$top ; $j--) { 
                $matrix[$j][$left]=$i++;
            }
            $left++;
            }
            $direction=($direction+1)%4;
        
        }
        return $matrix;
    }
}

File: 0059-spiral-matrix-ii/0059-spiral-matrix-ii.rs
impl Solution {
    pub fn generate_matrix(n: i32) -> Vec<Vec<i32>> {
        let n = n as usize;
        let mut res = vec![vec![0; n]; n];
        let mut count = 0;
        let (mut top, mut bottom, mut left, mut right) = (0, n - 1, 0, n - 1);
        
        while top <= bottom && left <= right {
            for i in left..=right {
                count += 1;
                res[top][i] = count;
            }
            top += 1;
            
            for i in top..=bottom {
                count += 1;
                res[i][right] = count;
            }
            right -= 1;
            if right == std::usize::MAX {
                break;
            }
            
            for i in (left..=right).rev() {
                if top > bottom {
                    continue;
                }
                count += 1;
                res[bottom][i] = count;
            }
            bottom -= 1;
            if bottom == std::usize::MAX {
                break;
            }
            
            for i in (top..=bottom).rev() {
                if left > right {
                    continue;
                }
                count += 1;
                res[i][left] = count;
            }
            left += 1;
        }
        res
    }
}